### PR TITLE
Add clear notifications to status

### PR DESCRIPTION
**Pull request by michealmikeyb**
Add a button in the menu to clear the notifications for a specific status to satisfy https://github.com/mastodon/mastodon/issues/22860 . Added an API endpoint at /api/v1/statuses/<STATUS_ID>/clear_notifications that the button will trigger that will delete all notifications from favourites, reblogs and replies to a status.
image

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -443,6 +443,7 @@
 
           resource :mute, only: :create
           post :unmute, to: 'mutes#destroy'
+          post :clear_notifications, to: 'mutes#clear_notifications'
```
Comment (ClearlyClaire) [inline config/routes.rb:446]: Putting that in the `MutesController` sound a bit weird tbh, but I'm not sure where it would fit best.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
Hi, thank you for your contribution!

I like the idea, but it seems to me it would not immediately reflect in the UI, which may be confusing. Therefore, I think a handler for `STATUS_CLEAR_NOTIFICATIONS_SUCCESS` should be added to `app/javascript/mastodon/reducers/notifications.js`.

Furthermore, you're adding a new API endpoint, please document it in the PR's description!

**Comment by michealmikeyb** [comment]
added the reducer and updated the PR description

File: config/routes.rb (hunk 1 above)
Comment (michealmikeyb) [reply config/routes.rb:446]: yeah same, felt it was sort of related to muting and didn't think it was enough for its own controller

**Comment by michealmikeyb** [comment]
@ClearlyClaire can i get another review, added the reducer

**Comment by michealmikeyb** [comment]
@ClearlyClaire i merged in master a bit ago is there anything else I should do?

File: app/javascript/mastodon/reducers/notifications.js [hunk 2]
```diff
@@ -175,6 +178,15 @@ const filterNotifications = (state, accountIds, type) => {
   return state.update('items', helper).update('pendingItems', helper);
 };
 
+const filterNotificationsByStatus = (state, statusId) => {
+  const helper = list => {
+    list = list.filterNot(item => item !== null && (item.get('type') === 'favourite' && item.get('status') === statusId));
+    list = list.filterNot(item => item !== null && (item.get('type') === 'reblog' && item.get('status') === statusId));
+    return list;
```
Comment (nschonni) [inline app/javascript/mastodon/reducers/notifications.js:185]:
```suggestion
    return list;
```

File: app/javascript/mastodon/actions/statuses.js [hunk 3]
```diff
@@ -230,6 +234,17 @@ export function muteStatus(id) {
...
+    dispatch(clearStatusNotificationRequest(id));
+
+    api(getState).post(`/api/v1/statuses/${id}/clear_notifications`).then(() => {
+      dispatch(clearStatusNotificationSuccess(id));
+    }).catch(error => {
+      dispatch(clearStatusNotificationFail(id, error));
+    });
+  };
+}
 export function muteStatusRequest(id) {
```
Comment (nschonni) [inline app/javascript/mastodon/actions/statuses.js:248]:
```suggestion

export function muteStatusRequest(id) {
```

File: app/controllers/api/v1/statuses/mutes_controller.rb [hunk 4]
```diff
@@ -22,6 +22,16 @@ def destroy
     render json: @status, serializer: REST::StatusSerializer
   end
 
+  def clear_notifications
+    current_account.notifications.where(activity_type: 'Mention').joins(mention: :status) \
+                   .where(status: { in_reply_to_id: params[:status_id] }).destroy_all
+    current_account.notifications.where(activity_type: 'Status').joins(:status) \
+                   .where(status: { reblog_of_id: params[:status_id] }).destroy_all
+    current_account.notifications.where(activity_type: 'Favourite').joins(:favourite) \
+                   .where(favourite: { status_id: params[:status_id] }).destroy_all
```
Comment (ClearlyClaire) [inline app/controllers/api/v1/statuses/mutes_controller.rb:31]: Please remove the extraneous `\` characters.

```suggestion
    current_account.notifications.where(activity_type: 'Mention').joins(mention: :status)
                   .where(status: { in_reply_to_id: params[:status_id] }).destroy_all
    current_account.notifications.where(activity_type: 'Status').joins(:status)
                   .where(status: { reblog_of_id: params[:status_id] }).destroy_all
    current_account.notifications.where(activity_type: 'Favourite').joins(:favourite)
                   .where(favourite: { status_id: params[:status_id] }).destroy_all
```

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
Apart from the inline comments, and the fact that the new translatable string is missing from `app/javascript/mastodon/locales/defaultMessages.json` and `app/javascript/mastodon/locales/en.json`, this seems fine to me.

It would also be good if you could rebase on main, as the tooling around testing and linting has changed quite a bit since you opened your PR (it is ok to force-push, and it is ok to squash your commits if that is easier for you).

The API change seems fine to me but I'd like @Gargron's opinion on this. Same on the UI change.

**Comment by michealmikeyb** [comment]
closing to try and re-evaluate diff

**Comment by michealmikeyb** [comment]
re opening

**Comment by nschonni** [comment]
You may want to create a copy of your branch, then try to rebase it with `it pull --rebase upstream main`, then force push

**Comment by michealmikeyb** [comment]
> You may want to create a copy of your branch, then try to rebase it with `it pull --rebase upstream main`, then force push

Thanks, that worked

**Comment by michealmikeyb** [comment]
@ClearlyClaire added the locales and rebased

**Comment by ClearlyClaire** [comment]
It looks like the translation string has been inserted in the wrong place.

You can run `yarn install && yarn build:development && yarn manage:translations en` to fix it automatically.

There's also a warning about a dependency that can be fixed by rebasing on `main` again.

I still want Gargron's input on the UI and API changes, so I'll request a review from him when the remaining issues are fixed, but otherwise it looks good to me.

File: app/javascript/mastodon/locales/en.json [hunk 5]
```diff
@@ -366,6 +366,8 @@
   "load_pending": "{count, plural, one {# new item} other {# new items}}",
   "loading_indicator.label": "Loading...",
   "media_gallery.toggle_visible": "{number, plural, one {Hide image} other {Hide images}}",
+  "missing_indicator.label": "Not found",
+  "missing_indicator.sublabel": "This resource could not be found",
```
Comment (ClearlyClaire) [inline app/javascript/mastodon/locales/en.json:370]: These seem to have been mistakenly re-added for some reason.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]

File: app/javascript/mastodon/locales/en.json (hunk 5 above)
Comment (nschonni) [reply app/javascript/mastodon/locales/en.json:370]: I had a similar thing happen on another PR, not sure where these are coming from. I had rebuilt first but the reappeared for me too, but not on CI

Comment (ClearlyClaire) [reply app/javascript/mastodon/locales/en.json:370]: Maybe leftover from `build/messages`?

Comment (nschonni) [reply app/javascript/mastodon/locales/en.json:370]: Looks like I needed to clean build and run `RAILS_ENV=development ./bin/rails assets:precompile` again before the other Yarn calls to clear those

Comment (michealmikeyb) [reply app/javascript/mastodon/locales/en.json]: removed it manually, when I tried to do it with build it kept making messages for some changes I had on a different branch

**Comment by mjankowski** [comment]
Closing due to lack of activity/rebase (~1 year since last activity) ... that said, it seems like this was at least possibly on road to merge.

If you want to nudge forward, can you rebase/reopen?

**Pull request closed without merging**